gdal/ogr/ogrsf_frmts/mssqlspatial/ogrmssqlgeometrywriter.cpp

```diff
@@ -203,41 +232,144 @@ void OGRMSSQLGeometryWriter::WritePoint(double x, double y)
 
 void OGRMSSQLGeometryWriter::WritePoint(double x, double y, double z)
 {
-    if (nColType == MSSQLCOLTYPE_GEOGRAPHY)
+    WriteZ(iPoint, z);
+    WritePoint(x, y);
+}
+
+void OGRMSSQLGeometryWriter::WritePoint(double x, double y, double z, double m)
+{
+    WriteZ(iPoint, z);
+    WriteM(iPoint, m);
+    WritePoint(x, y);
+}
+
+/************************************************************************/
+/*                         WriteSimpleCurve()                           */
+/************************************************************************/
+
+void OGRMSSQLGeometryWriter::WriteSimpleCurve(OGRSimpleCurve* poGeom, int iStartIndex, int nCount)
+{
+    if ((chProps & SP_HASZVALUES) && (chProps & SP_HASMVALUES))
     {
-        WriteY(iPoint, x);
-        WriteX(iPoint, y);
+        for (int i = iStartIndex; i < iStartIndex + nCount; i++)
+            WritePoint(poGeom->getX(i), poGeom->getY(i), poGeom->getZ(i), poGeom->getM(i));
+    }
+    else if (chProps & SP_HASZVALUES)
+    {
+        for (int i = iStartIndex; i < iStartIndex + nCount; i++)
+            WritePoint(poGeom->getX(i), poGeom->getY(i), poGeom->getZ(i));
+    }
+    else if (chProps & SP_HASMVALUES)
+    {
+        for (int i = iStartIndex; i < iStartIndex + nCount; i++)
+            WritePoint(poGeom->getX(i), poGeom->getY(i), poGeom->getM(i));
     }
     else
     {
-        WriteX(iPoint, x);
-        WriteY(iPoint, y);
+        for (int i = iStartIndex; i < iStartIndex + nCount; i++)
+            WritePoint(poGeom->getX(i), poGeom->getY(i));
     }
-    WriteZ(iPoint, z);
-    ++iPoint;
+}
+
+void OGRMSSQLGeometryWriter::WriteSimpleCurve(OGRSimpleCurve* poGeom, int iStartIndex)
+{
+    WriteSimpleCurve(poGeom, iStartIndex, poGeom->getNumPoints() - iStartIndex);
+}
+
+
+void OGRMSSQLGeometryWriter::WriteSimpleCurve(OGRSimpleCurve* poGeom)
+{
+    WriteSimpleCurve(poGeom, 0, poGeom->getNumPoints());
 }
 
 /************************************************************************/
-/*                         WriteLineString()                            */
+/*                         WriteCompoundCurve()                         */
 /************************************************************************/
 
-void OGRMSSQLGeometryWriter::WriteLineString(OGRLineString* poGeom)
+void OGRMSSQLGeometryWriter::WriteCompoundCurve(OGRCompoundCurve* poGeom)
 {
-    int i;
-    /* write figure */
-    WriteByte(FigureAttribute(iFigure), 0x01);
+    OGRSimpleCurve* poSubGeom;
+    WriteByte(FigureAttribute(iFigure), FA_CURVE);
     WriteInt32(PointOffset(iFigure), iPoint);
-    if (chProps & SP_HASZVALUES)
+    for (int i = 0; i < poGeom->getNumCurves(); i++)
     {
-        for (i = 0; i < poGeom->getNumPoints(); i++)
-            WritePoint(poGeom->getX(i), poGeom->getY(i), poGeom->getZ(i));
+        poSubGeom = poGeom->getCurve(i)->toSimpleCurve();
+        switch (poSubGeom->getGeometryType())
```

Comment:
minor: you coud use          switch (wkbFlatten(poSubGeom->getGeometryType())) and use only the 2D version of the wkb constants in the case 
This holds for similar switch() in other files

Reply:
@rouault Do you think if the code is more readable in that case? We should also enumerate the M/ZM variants however.

Reply:
> We should also enumerate the M/ZM variants however.

no, the code would become just
```
switch (wkbFlatten(poSubGeom->getGeometryType()))
{
        case wkbLineString:
            // do stuff
            break;
        case wkbCircularString:
            // do stuff
            break;
        default:
            break;
}
```
Looks better to me